### PR TITLE
Config file support

### PR DESCRIPTION
**Pull request by mslosarek**
Added support for modifying the config file

### CODE REVIEW

File: switcher.js [hunk 1]
```diff
@@ -133,10 +141,29 @@ class Switcher {
     return 'bad';
   }
 
+  /**
+   * Checks that the config file has valid data
+   *
+   * @returns {String} - Status of the file's contents
+   */
+  _checkConfigFile() {
+    if (!fs.existsSync(`${this.awsDir}/config`)) {
```
Comment (devtanc) [inline switcher.js]: This existence check should really be part of the `_checkDirectory` function, which should also be modified to be something like:

```js
_checkDirectory() {
  if (!fs.existsSync(`${this.awsDir}/config`)) throw new Error(`No config file found in ${this.awsDir}`)
  if (!fs.existsSync(`${this.awsDir}/credentials`)) throw new Error(`No credentials file found in ${this.awsDir}`)
}
```

Then simply called in the constructor replacing lines #26-29

```js
if (!this._checkDirectory(this.awsDir)) {
  throw new Error(`No credentials file found in ${this.awsDir}`);
}
```

File: switcher.js [hunk 2]
```diff
@@ -133,10 +141,29 @@ class Switcher {
...
+      return 'bad';
+    }
+    const contents = fs.readFileSync(`${this.awsDir}/config`, 'utf8');
+    if (contents === '') {
+      return 'empty';
+    }
+    const regexMatch = contents.match(/\[.*\]\n* *output *= *.*\n *region *= *.*/);
```
Comment (devtanc) [inline switcher.js]: According to [this](https://docs.aws.amazon.com/cli/latest/topic/config-vars.html) the config file may have nested values, and additional values beyond just the region and output, like so:

```
[profile testing]
aws_access_key_id = foo
aws_secret_access_key = bar
region = us-west-2
s3 =
  max_concurrent_requests=10
  max_queue_size=1000
```

This would cause the regex here to [fail to match](https://regex101.com/r/LkHUmp/1) the file, even though the format of the file is good.

File: switcher.js [hunk 3]
```diff
@@ -201,33 +228,141 @@ class Switcher {
...
+            * Which means it's either an AWS profile ID or a configuration which are are formatted like:
+            *  output = OUTPUT
+            *  region = REGION
+            * With spaces in between the '=' and the key/value
+            */
+
+            // Split on '=' and then remove any spaces
+            // [0] === 'output'  || [0] === 'region'
+            // [1] === OUTPUT    || [1] === REGION
+            const [key, value] = curr.split('=').map(section => section.replace(' ', ''));
```
Comment (devtanc) [inline switcher.js]: If [this format](https://docs.aws.amazon.com/cli/latest/topic/config-vars.html) is possible

```
[profile testing]
aws_access_key_id = foo
aws_secret_access_key = bar
region = us-west-2
s3 =
  max_concurrent_requests=10
  max_queue_size=1000
```

then we're going to have issues here with storing values like this `s3` one. Need to figure out a way to handle those edge cases.

File: switcher.js [hunk 4]
```diff
@@ -201,33 +228,141 @@ class Switcher {
...
 
+  /**
+   * Parses the config file into an Array of profile objects,
+   * appends the config settings to the existing profile object
+   *
+   * @param {array} Profile objects - Existing array or profile objects
+   * @returns {Promise.<Array>} - Array of profile objects
+   * @reject {Error} - fs.readFile error
+   */
+  _getConfig(profileData) {
```
Comment (devtanc) [inline switcher.js]: This function is nearly identical to `_getCredentials`, which is fine. It's just that a lot of the logic in here should probably be extracted so that it can be shared between the two functions and not duplicated.

File: switcher.js [hunk 5]
```diff
@@ -201,33 +228,141 @@ class Switcher {
...
    * @param {Object[]} profile - Array of profile objects
    * @param {String} profile.name                   - The profile's name (no square brackets)
    * @param {String} profile.aws_access_key_id      - The profile's access key ID (from AWS)
    * @param {String} profile.aws_secret_access_key  - The profile's secret access key (from AWS)
+   * @param {String} profile.output  - The profile's output type
+   * @param {String} profile.region  - The profile's region location
    * @returns {Promise.<undefined>} - Returns an empty resolved Promise
    * @reject {Error} - Error writing the credentials file
    */
   _writeBackCredentialsData(data) {
```
Comment (devtanc) [inline switcher.js]: I'd rather that there was something like a `_writeFiles` function that was called, which in turn called both `_writeBackCredentialsData` and a `_writeBackConfigFileData` function and handled any errors, that way the config file logic and the credentials file logic stay separate and clean.

File: test/mocha.js [hunk 6]
```diff
@@ -236,4 +270,53 @@ describe('Switcher test suite', () => {
       });
     });
   });
+
```
Comment (devtanc) [inline test/mocha.js]: Need some tests here that cover those edge cases that I previously mentioned, just to make sure that we're not screwing up anybody's file format.

File: switcher.js [hunk 7]
```diff
@@ -35,7 +35,16 @@ class Switcher {
...
 
-    this.profileData = this._getCredentials(this.awsDir);
+    const configResult = this._checkConfigFile(this.awsDir);
+    if (configResult === 'good') {
+      this.processConfig = true;
+    } else {
+      this.processConfig = false;
+    }
+
+    this.profileData = this._getCredentials().then(profileData => {
```
Comment (devtanc) [inline switcher.js]: It would probably be better to combine these differently, since it's possible that there would be collisions in the keys between the credentials file and the config file. Maybe having this data as:

```js
{
  profileData: {},
  configData: {},
}
```

or maybe just having a separate property on `this` to hold the data from the config file. I think attaching a new `this` property might end up cleaner, and might allow better separation of the logic where necessary.

**Comment by HuJake** [comment]
Config file support can merge to master?

**Comment by devtanc** [comment]
No, not currently @HuJake. Those code issues need to be fixed or we'll have bugs introduced that would be a problem.
